File: src/Cocoar.SignalARRR.Typescript/src/harrr-connection.ts
import * as signalR from '@microsoft/signalr';
import { ClientRequestMessage } from './models/client-request-message.js';
import { ServerRequestMessage } from './models/server-request-message.js';
import { asCancellationTokenReference } from './models/cancellation-token-reference.js';
import { isStreamReference, resolveStreamReference } from './models/stream-reference.js';
import { parseHARRRError } from './models/harrr-error.js';
import { HARRRConnectionOptions } from './harrr-connection-options.js';
import { CancellationManager } from './cancellation-manager.js';

export class HARRRConnection {
  private _hubConnection: signalR.HubConnection;
  // SignalR's contract is `() => string | Promise<string>`, and every OAuth-backed application
  // returns the promise. Typing the field as synchronous did not make it so: the promise itself was
  // serialised into `ClientRequestMessage.Authorization`, arrived as `{}`, and the server — where
  // the field is a `string` — failed to bind the whole message. Every invoke, send and stream on
  // that connection died with "Error binding arguments", while server-to-client calls kept working
  // because they carry no such message. Both .NET clients and the Swift client await the token on
  // every send path; this one now does too.
  private _accessTokenFactory: () => string | Promise<string> = () => '';
  private _serverRequestHandlers = new Map<string, (...args: unknown[]) => unknown>();
  private _serverStreamHandlers = new Map<string, (...args: unknown[]) => AsyncIterable<unknown>>();
  private _cancellationManager = new CancellationManager();

  public get baseUrl(): string {
    return this._hubConnection.baseUrl;
  }

  public set baseUrl(value: string) {
    this._hubConnection.baseUrl = value;
  }

  public get connectionId(): string | null {
    return this._hubConnection.connectionId;
  }

  public get state(): signalR.HubConnectionState {
    return this._hubConnection.state;
  }

  public get serverTimeoutInMilliseconds(): number {
    return this._hubConnection.serverTimeoutInMilliseconds;
  }

  public set serverTimeoutInMilliseconds(value: number) {
    this._hubConnection.serverTimeoutInMilliseconds = value;
  }

  public get keepAliveIntervalInMilliseconds(): number {
    return this._hubConnection.keepAliveIntervalInMilliseconds;
  }

  public set keepAliveIntervalInMilliseconds(value: number) {
    this._hubConnection.keepAliveIntervalInMilliseconds = value;
  }

  constructor(hubConnection: signalR.HubConnection, _options?: HARRRConnectionOptions) {
    this._hubConnection = hubConnection;

    const conn = (hubConnection as unknown as Record<string, unknown>)['connection'] as Record<string, unknown> | undefined;
    const factory =
      (conn?.['_options'] as Record<string, unknown> | undefined)?.['accessTokenFactory'] ??
      conn?.['_accessTokenFactory'];
    if (typeof factory === 'function') {
      this._accessTokenFactory = factory as () => string | Promise<string>;
    }

    // Native client results — return values are sent back to the server automatically by SignalR.
    // A promise is fine here without awaiting it ourselves: SignalR awaits the return value of a
    // client-result handler before completing the invocation.
    this._hubConnection.on('ChallengeAuthentication', (req: ServerRequestMessage) => {
      return this._accessTokenFactory();
    });

    this._hubConnection.on('InvokeServerRequest', async (req: ServerRequestMessage) => {
      // If StreamId is present, stream results back instead of returning a single value
      if (req.StreamId) {
        await this._streamBackToServer(req, req.StreamId);
        return undefined;
      }
      const result = await this._dispatchServerMethod(req);

      // If the result is binary data (Blob, ArrayBuffer, Buffer), upload via HTTP
      // and return a StreamReference instead
      if (result instanceof Blob || result instanceof ArrayBuffer ||
          isNodeBuffer(result)) {
        return await this._uploadAndReturnReference(result);
      }

      return result;
    });

    this._hubConnection.on('InvokeServerMessage', async (req: ServerRequestMessage) => {
      try {
        if (req.StreamId) {
          await this._streamBackToServer(req, req.StreamId);
        } else {
          await this._dispatchServerMethod(req);
        }
      } catch (err) {
        console.error(`[SignalARRR] Failed to handle server message '${req.Method}':`, err);
      }
    });

    this._hubConnection.on('CancelTokenFromServer', (req: ServerRequestMessage) => {
      if (req.CancellationGuid) {
        this._cancellationManager.cancel(req.CancellationGuid);
      }
    });
  }

  private async _dispatchServerMethod(req: ServerRequestMessage): Promise<unknown> {
    const handler = this._serverRequestHandlers.get(req.Method);
    if (!handler) return undefined;

    const args = await this._prepareArgs(req);
    return await handler(...args);
  }

  private async _prepareArgs(req: ServerRequestMessage): Promise<unknown[]> {
    const args: unknown[] = [];
    for (const arg of req.Arguments ?? []) {
      const cancellationRef = asCancellationTokenReference(arg);
      if (cancellationRef) {
        // Keyed on the reference's own id, not on the request's CancellationGuid. Those are two
        // different things: the guid cancels the call, the reference cancels this one parameter.
        // Registering under the guid meant a cancellation aimed at a parameter never found its
        // token here, and two token parameters could not be cancelled apart.
        args.push(this._cancellationManager.create(cancellationRef.Id));
      } else if (isStreamReference(arg)) {
        // Download the stream data via HTTP and pass as ArrayBuffer
        args.push(await resolveStreamReference(arg));
      } else {
        args.push(arg);
      }
    }
    return args;
  }

  private async _streamBackToServer(req: ServerRequestMessage, streamId: string): Promise<void> {
    const args = await this._prepareArgs(req);

    try {
      // Try stream handler first
      const streamHandler = this._serverStreamHandlers.get(req.Method);
      if (streamHandler) {
        const stream = streamHandler(...args);
        for await (const item of stream) {
          await this._hubConnection.send('StreamItemToServer', streamId, item);
        }
        await this._hubConnection.send('StreamCompleteToServer', streamId, null);
        return;
      }

      // Fall back to regular handler — send single result as one item
      const handler = this._serverRequestHandlers.get(req.Method);
      if (handler) {
        const result = await handler(...args);
        if (result != null) {
          // Check if result is async iterable
          if (typeof result === 'object' && Symbol.asyncIterator in (result as object)) {
            for await (const item of result as AsyncIterable<unknown>) {
              await this._hubConnection.send('StreamItemToServer', streamId, item);
            }
          } else {
            await this._hubConnection.send('StreamItemToServer', streamId, result);
          }
        }
        await this._hubConnection.send('StreamCompleteToServer', streamId, null);
      } else {
        await this._hubConnection.send('StreamCompleteToServer', streamId, null);
      }
    } catch (err) {
      await this._hubConnection.send('StreamCompleteToServer', streamId, String(err));
    }
  }

  private async _uploadAndReturnReference(data: Blob | ArrayBuffer | Uint8Array | unknown): Promise<{ Uri: string }> {
    // Request an upload URL from the server
    const uploadUrl = await this._hubConnection.invoke<string>('RequestUploadSlot');

    // Upload the data via HTTP POST
    let body: BodyInit;
    if (data instanceof Blob) {
      body = data;
    } else if (data instanceof ArrayBuffer || data instanceof Uint8Array) {
      body = data as BodyInit;
    } else {
      body = String(data);
    }
    await fetch(uploadUrl, {
      method: 'POST',
      headers: { 'Content-Type': 'application/octet-stream' },
      body,
    });

    return { Uri: uploadUrl };
  }

  /** Prepare outgoing arguments — upload binary data and replace with StreamReferences. */
  private async _prepareOutgoingArgs(args: unknown[]): Promise<unknown[]> {
    let hasStream = false;
    for (const arg of args) {
      if (arg instanceof Blob || arg instanceof ArrayBuffer ||
          isNodeBuffer(arg)) {
        hasStream = true;
        break;
      }
    }
    if (!hasStream) return args;

    const result: unknown[] = [];
    for (const arg of args) {
      if (arg instanceof Blob || arg instanceof ArrayBuffer ||
          isNodeBuffer(arg)) {
        result.push(await this._uploadAndReturnReference(arg));
      } else {
        result.push(arg);
      }
    }
    return result;
  }

  public start(): Promise<void> {
    return this._hubConnection.start();
  }

  public stop(): Promise<void> {
    return this._hubConnection.stop();
  }

  public onClose(callback: (error?: Error) => void): void {
    this._hubConnection.onclose(callback);
  }

  public onReconnecting(callback: (error?: Error) => void): void {
    this._hubConnection.onreconnecting(callback);
  }

  public onReconnected(callback: (connectionId?: string) => void): void {
    this._hubConnection.onreconnected(callback);
  }

  /** Resolves the token factory, sync or async, to the string the message field expects. */
  private async _resolveAuthorization(): Promise<string> {
    return (await this._accessTokenFactory()) ?? '';
  }

  private async _buildRequest(methodName: string, args: unknown[]): Promise<ClientRequestMessage> {
    const preparedArgs = await this._prepareOutgoingArgs(args);
    return {
      Method: methodName,
      Arguments: preparedArgs,
      Authorization: await this._resolveAuthorization(),
    };
  }

  public async invoke<T>(methodName: string, ...args: unknown[]): Promise<T> {
    const msg = await this._buildRequest(methodName, args);
    return this._hubConnection
      .invoke<T>('InvokeMessageResult', msg)
      .catch(err => Promise.reject(this._extractException(err)));
  }

  public async send(methodName: string, ...args: unknown[]): Promise<void> {
    const msg = await this._buildRequest(methodName, args);
    return this._hubConnection.send('SendMessage', msg);
  }

  /**
   * `IStreamResult` has to be handed back synchronously, but building the message is asynchronous —
   * the token has to be awaited, and binary arguments have to be uploaded first, which this path
   * used to skip entirely. The result is therefore deferred: it starts the real stream once the
   * message is ready and forwards to whoever subscribed in the meantime.
   */
  public stream<T>(methodName: string, ...args: unknown[]): signalR.IStreamResult<T> {
    const pending = this._buildRequest(methodName, args).then(msg =>
      this._hubConnection.stream<T>('StreamMessage', msg),
    );
    return new DeferredStreamResult<T>(pending);
  }

  public on(methodName: string, newMethod: (...args: any[]) => void): void {
    this._hubConnection.on(methodName, newMethod);
  }

  /** Register a handler for server-to-client method calls (single return value). */
  public onServerMethod(methodName: string, func: (...args: unknown[]) => unknown): this {
    this._serverRequestHandlers.set(methodName, func);
    return this;
  }

  /** Register a handler for server-to-client streaming calls (returns AsyncIterable). */
  public onServerStreamMethod(methodName: string, func: (...args: unknown[]) => AsyncIterable<unknown>): this {
    this._serverStreamHandlers.set(methodName, func);
    return this;
  }

  public off(methodName: string): void;
  public off(methodName: string, method: (...args: any[]) => void): void;
  public off(methodName: string, method?: (...args: any[]) => void): void {
    if (!method) {
      this._hubConnection.off(methodName);
    } else {
      this._hubConnection.off(methodName, method);
    }
  }

  public asSignalRHubConnection(): signalR.HubConnection {
    return this._hubConnection;
  }

  public static create(
    hubConnection: signalR.HubConnection | ((builder: signalR.HubConnectionBuilder) => void),
    options?: HARRRConnectionOptions,
  ): HARRRConnection {
    if (hubConnection instanceof Function) {
      const builder = new signalR.HubConnectionBuilder();
      hubConnection(builder);
      return new HARRRConnection(builder.build(), options);
    }
    return new HARRRConnection(hubConnection, options);
  }

  private _extractException(error: unknown): { type: string; message: string } {
    const parsed = parseHARRRError(error);
    return { type: parsed.Type, message: parsed.Message };
  }
}

/**
 * An `IStreamResult` whose underlying stream is not open yet.
 *
 * Subscribers registered before the stream exists are attached as soon as it does; a failure while
 * preparing the message — a rejecting token factory, a failed upload — reaches them through
 * `error()` rather than being thrown out of `stream()`, which is where a caller can observe it.
 * Disposing before the stream opens means it is never subscribed to at all.
 */
class DeferredStreamResult<T> implements signalR.IStreamResult<T> {
  constructor(private readonly _pending: Promise<signalR.IStreamResult<T>>) {
    // The rejection is delivered to subscribers below. This keeps it from also counting as an
    // unhandled rejection when nobody subscribes, which in Node terminates the process.
    this._pending.catch(() => undefined);
  }

  public subscribe(subscriber: signalR.IStreamSubscriber<T>): signalR.ISubscription<T> {
    let inner: signalR.ISubscription<T> | undefined;
    let disposed = false;

    this._pending.then(
      stream => {
        if (disposed) return;
        inner = stream.subscribe(subscriber);
      },
      err => {
        if (disposed) return;
        subscriber.error(err);
      },
    );

    return {
      dispose: () => {
        disposed = true;
        inner?.dispose();
      },
    };
  }
}

/** Runtime check for Node.js Buffer without importing @types/node */
function isNodeBuffer(value: unknown): boolean {
  return typeof globalThis !== 'undefined' &&
    typeof (globalThis as Record<string, unknown>)['Buffer'] === 'function' &&
    typeof ((globalThis as Record<string, unknown>)['Buffer'] as Record<string, unknown>)['isBuffer'] === 'function' &&
    ((globalThis as Record<string, unknown>)['Buffer'] as { isBuffer: (v: unknown) => boolean }).isBuffer(value);
}

File: src/Cocoar.SignalARRR.Kotlin/signalarrr/src/main/kotlin/dev/cocoar/signalarrr/HARRRConnection.kt
package dev.cocoar.signalarrr

import dev.cocoar.signalarrr.serialization.JsonValues
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineName
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonPrimitive
import okhttp3.OkHttpClient
import java.util.concurrent.ConcurrentHashMap
import kotlin.time.Duration

/** Handles a server-to-client call; the returned value is the result (ignored for fire-and-forget calls). */
public typealias ServerMethodHandler = suspend (args: ServerMethodArgs) -> Any?

/** Handles a server-to-client streaming call; each emitted item is sent back to the server. */
public typealias ServerStreamMethodHandler = suspend (args: ServerMethodArgs) -> Flow<Any?>

/**
 * Options of a [HARRRConnection]; the [SignalRClientOptions] part configures the connection underneath.
 *
 * A connection has two credentials, checked by different things. The **connection credential**
 * authenticates negotiate and the transport and is what `[Authorize]` on the hub class checks. The
 * **message credential** travels with every message, answers token challenges and authorises file
 * transfers, and is what `[Authorize]` on a method or a `ServerMethods` class checks. The options
 * are named the same in every SignalARRR client, and nothing is coupled implicitly: setting one of
 * the two sets only that one.
 */
public class HARRRConnectionOptions : SignalRClientOptions() {
    /**
     * One credential for the connection and for every message — the common case. A
     * [connectionCredential] or [messageCredential] set alongside it takes its part over.
     */
    public var credential: (suspend () -> String?)? = null

    /**
     * Authenticates the connection: the negotiate request and the transport. Called on every
     * connect and reconnect. Takes the place of [SignalRClientOptions.accessTokenProvider]; set
     * only one of the two.
     */
    public var connectionCredential: (suspend () -> String?)? = null

    /**
     * Authenticates each message: travels as `ClientRequestMessage.Authorization`, answers token
     * challenges, and authorises file-transfer requests. Called per use, so a refreshed token is
     * picked up.
     */
    public var messageCredential: (suspend () -> String?)? = null

    /**
     * The message credential under its former name. Unlike [messageCredential], it also
     * authenticates the connection when [SignalRClientOptions.accessTokenProvider] is left `null` —
     * the implicit coupling the new options no longer have.
     */
    @Deprecated("Use messageCredential, or credential for one credential covering the connection and every message.")
    public var messageAccessTokenProvider: (suspend () -> String?)? = null

    /** The JSON configuration used to encode arguments and decode results. */
    public var json: Json = JsonValues.defaultJson
}

/**
 * The SignalARRR client: typed bidirectional RPC over a [SignalRClient].
 *
 * `invoke`, `send` and `stream` call the server; [onServerMethod] and friends answer calls from
 * the server. Token challenges, cancellation propagation and HTTP stream references are handled
 * underneath.
 */
public class HARRRConnection private constructor(
    /** The SignalR connection underneath — for raw hub methods and connection events. */
    public val client: SignalRClient,
    private val messageAccessTokenProvider: suspend () -> String?,
    /** The JSON configuration of this connection. */
    public val json: Json,
    httpClient: OkHttpClient,
    public val serverTimeout: Duration,
    public val keepAliveInterval: Duration,
    public val handshakeTimeout: Duration,
) {
    /** Cancellation ids of running server-to-client handlers. */
    public val cancellationManager: CancellationManager = CancellationManager()

    private val streamResolver = StreamReferenceResolver(httpClient)
    private val serverHandlers = ConcurrentHashMap<String, ServerMethodHandler>()
    private val serverStreamHandlers = ConcurrentHashMap<String, ServerStreamMethodHandler>()
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO + CoroutineName("HARRRConnection"))

    /** Observes every server request envelope before it is dispatched. */
    public var onServerRequestMessageReceived: ((ServerRequestMessage) -> Unit)? = null

    public val state: HubConnectionState get() = client.state
    public val connectionId: String? get() = client.connectionId

    public companion object {
        /**
         * Creates a connection to [url]. Credentials come from [HARRRConnectionOptions.credential],
         * [HARRRConnectionOptions.connectionCredential] and [HARRRConnectionOptions.messageCredential].
         *
         * @throws IllegalArgumentException when a credential is set in two places — the former
         *   `messageAccessTokenProvider` next to the new options, or `accessTokenProvider` next to
         *   `connectionCredential`/`credential`.
         */
        public fun create(url: String, configure: HARRRConnectionOptions.() -> Unit = {}): HARRRConnection {
            val options = HARRRConnectionOptions().apply(configure)
            @Suppress("DEPRECATION")
            val legacyMessage = options.messageAccessTokenProvider
            val usesNewOptions = options.credential != null || options.connectionCredential != null || options.messageCredential != null
            val messageProvider = if (usesNewOptions) {
                require(legacyMessage == null) {
                    "messageAccessTokenProvider is the former name of messageCredential; set only messageCredential (or credential)."
                }
                val connection = options.connectionCredential ?: options.credential
                if (connection != null) {
                    require(options.accessTokenProvider == null) {
                        "The connection credential is set twice: through accessTokenProvider and through connectionCredential/credential. Set only one of them."
                    }
                    options.accessTokenProvider = connection
                }
                options.messageCredential ?: options.credential
            } else {
                // The former option keeps its former behaviour: it also covers the connection.
                if (options.accessTokenProvider == null) options.accessTokenProvider = legacyMessage
                legacyMessage
            }
            val client = SignalRClient(url, options)
            return create(client, messageProvider, options.json, options)
        }

        /** Wraps an existing [SignalRClient]. */
        public fun create(
            client: SignalRClient,
            messageAccessTokenProvider: (suspend () -> String?)? = null,
            json: Json = JsonValues.defaultJson,
            options: SignalRClientOptions = SignalRClientOptions(),
        ): HARRRConnection {
            val connection = HARRRConnection(
                client = client,
                messageAccessTokenProvider = messageAccessTokenProvider ?: { null },
                json = json,
                httpClient = options.httpClient ?: OkHttpClient(),
                serverTimeout = options.serverTimeout,
                keepAliveInterval = options.keepAliveInterval,
                handshakeTimeout = options.handshakeTimeout,
            )
            connection.registerBuiltInHandlers()
            return connection
        }
    }

    // ---------------------------------------------------------------- lifecycle

    public suspend fun start(): Unit = client.start()

    public suspend fun stop(): Unit = client.stop()

    public fun onClosed(callback: suspend (error: Throwable?) -> Unit): Unit = client.onClosed(callback)
    public fun onReconnecting(callback: suspend (error: Throwable?) -> Unit): Unit = client.onReconnecting(callback)
    public fun onReconnected(callback: suspend (connectionId: String?) -> Unit): Unit = client.onReconnected(callback)

    // ---------------------------------------------------------------- client → server

    /** Calls [method] and decodes its result as [T]. */
    public suspend inline fun <reified T> invoke(method: String, vararg arguments: Any?, genericArguments: List<String> = emptyList()): T =
        JsonValues.decode(invokeCore(method, arguments.toList(), genericArguments), json)

    /** Calls [method] and returns the raw result. `ByteArray`/`File`/`InputStream` arguments are uploaded first. */
    public suspend fun invokeCore(method: String, arguments: List<Any?>, genericArguments: List<String> = emptyList()): JsonElement? {
        val request = buildClientRequest(method, arguments, genericArguments)
        return translateErrors {
            client.invoke(MethodNames.INVOKE_MESSAGE_RESULT, listOf(encodeRequest(request)))
        }
    }

    /** Fire-and-forget call of [method]. */
    public suspend fun send(method: String, vararg arguments: Any?, genericArguments: List<String> = emptyList()) {
        val request = buildClientRequest(method, arguments.toList(), genericArguments)
        translateErrors {
            client.send(MethodNames.SEND_MESSAGE, listOf(encodeRequest(request)))
        }
    }

    /** Opens a stream from [method] and decodes each item as [T]. Cancelling the collector cancels the stream. */
    public inline fun <reified T> stream(method: String, vararg arguments: Any?, genericArguments: List<String> = emptyList()): Flow<T> =
        streamCore(method, arguments.toList(), genericArguments).map { JsonValues.decode<T>(it, json) }

    public fun streamCore(method: String, arguments: List<Any?>, genericArguments: List<String> = emptyList()): Flow<JsonElement> = flow {
        val request = buildClientRequest(method, arguments, genericArguments)
        try {
            client.stream(MethodNames.STREAM_MESSAGE, listOf(encodeRequest(request))).collect { emit(it) }
        } catch (e: HubInvocationException) {
            throw translate(e)
        }
    }

    /** The typed proxy of a contract — pass the generated proxy's companion, e.g. `getTypedMethods(IChatHubProxy)`. */
    public fun <T : Any> getTypedMethods(factory: HubProxyFactory<T>): T = factory.create(this)

    // ---------------------------------------------------------------- server → client

    /**
     * Registers the handler of a SignalARRR server-to-client contract method. [name] is the wire
     * name, `Interface|Method` (see the contract wire names guide); it is matched exactly. Use
     * [on] instead for raw SignalR events sent with `Clients.X.SendAsync("name", ...)`.
     *
     * The handler runs in a coroutine that the server can cancel when the contract declares a
     * `CancellationToken`. Return the result, or `Unit` for void methods; a `ByteArray`, `File`
     * or `InputStream` result is uploaded and returned to the server as a stream reference.
     */
    public fun onServerMethod(name: String, handler: ServerMethodHandler) {
        serverHandlers[name] = handler
    }

    /** Registers the handler of a server-to-client streaming method (`IAsyncEnumerable<T>` on the contract). */
    public fun onServerStreamMethod(name: String, handler: ServerStreamMethodHandler) {
        serverStreamHandlers[name] = handler
    }

    public fun removeServerMethod(name: String) {
        serverHandlers.remove(name)
        serverStreamHandlers.remove(name)
    }

    /** Registers [handlers] under `prefix|name`. */
    public fun registerHandlers(prefix: String, handlers: Map<String, ServerMethodHandler>) {
        handlers.forEach { (name, handler) -> serverHandlers["$prefix|$name"] = handler }
    }

    /** Registers every handler of [handler] under its interface name. */
    public fun registerInterface(handler: ServerInterfaceHandler) {
        registerHandlers(handler.interfaceName, handler.handlers())
    }

    /**
     * Registers the handler of a raw SignalR event. The returned value becomes the client result.
     * SignalARRR contract calls do not arrive here — use [onServerMethod] for those.
     */
    public fun on(method: String, handler: suspend (args: ServerMethodArgs) -> Any?) {
        client.on(method) { arguments ->
            val result = handler(ServerMethodArgs(arguments, json))
            if (result == null || result == Unit) null else JsonValues.encode(result, json)
        }
    }

    public fun off(method: String): Unit = client.off(method)

    // ---------------------------------------------------------------- built-in wire handlers

    private fun registerBuiltInHandlers() {
        client.on(MethodNames.CHALLENGE_AUTHENTICATION) {
            val token = messageAccessTokenProvider() ?: ""
            JsonPrimitive(token)
        }

        client.on(MethodNames.INVOKE_SERVER_REQUEST) { arguments ->
            val request = decodeServerRequest(arguments)
            onServerRequestMessageReceived?.invoke(request)
            val result = dispatchServerMethod(request)
            encodeResult(result)
        }

        client.on(MethodNames.INVOKE_SERVER_MESSAGE) { arguments ->
            val request = decodeServerRequest(arguments)
            onServerRequestMessageReceived?.invoke(request)
            val streamId = request.streamId
            if (streamId != null) {
                scope.launch { streamBackToServer(request, streamId) }
            } else {
                try {
                    dispatchServerMethod(request)
                } catch (e: CancellationException) {
                    if (!currentCoroutineContext().isActive) throw e
                    client.logger.warning { "Server message '${request.method}' was cancelled: ${e.message}" }
                } catch (e: Throwable) {
                    client.logger.error(e) { "Failed to handle server message '${request.method}': ${e.message}" }
                }
            }
            null
        }

        client.on(MethodNames.CANCEL_TOKEN_FROM_SERVER) { arguments ->
            val request = decodeServerRequest(arguments)
            request.cancellationGuid?.let { cancellationManager.cancel(it) }
            null
        }

        client.onClosed { cancellationManager.cancelAll("Connection closed") }
        client.onReconnecting { cancellationManager.cancelAll("Connection lost") }
    }

    private fun decodeServerRequest(arguments: List<JsonElement>): ServerRequestMessage {
        val first = arguments.firstOrNull() ?: throw InvocationFailedException("No arguments in server request")
        return JsonValues.decode(first, ServerRequestMessage.serializer(), json)
    }

    /**
     * Runs the handler of [request] in a child coroutine registered with the cancellation
     * manager, so `CancelTokenFromServer` cancels exactly this call. The cancellation surfaces
     * here as a [ServerCancelledException] rather than cancelling the dispatching coroutine, so
     * the error can still be reported back to the server.
     */
    private suspend fun dispatchServerMethod(request: ServerRequestMessage): Any? {
        val handler = serverHandlers[request.method]
        if (handler == null) {
            client.logger.warning {
                "No server-method handler registered for '${request.method}'. Register it with " +
                    "onServerMethod(\"${request.method}\") { ... } — on(...) only handles raw SignalR events, " +
                    "not SignalARRR server→client contracts."
            }
            return null
        }
        return runHandler(request) { args -> handler(args) }
    }

    private suspend fun <R> runHandler(request: ServerRequestMessage, block: suspend (ServerMethodArgs) -> R): R = coroutineScope {
        val ids = mutableListOf<String>()
        val deferred = async {
            val args = buildHandlerArgs(request, ids, currentCoroutineContext()[kotlinx.coroutines.Job]!!)
            block(args)
        }
        request.cancellationGuid?.let { ids.add(it) }
        // The handler may not have registered its ids yet: register the guid now, the argument
        // ids as soon as they are known (buildHandlerArgs adds them).
        request.cancellationGuid?.let { cancellationManager.register(it, deferred) }
        try {
            deferred.await()
        } catch (e: CancellationException) {
            if (deferred.isCancelled && currentCoroutineContext().isActive) {
                val cause = generateSequence<Throwable>(e) { it.cause }.firstOrNull { it is ServerCancelledException }
                throw OperationCancelledException(cause?.message ?: e.message ?: "The operation was cancelled", e)
            }
            throw e
        } finally {
            ids.forEach { cancellationManager.remove(it) }
        }
    }

    private suspend fun buildHandlerArgs(request: ServerRequestMessage, ids: MutableList<String>, job: kotlinx.coroutines.Job): ServerMethodArgs {
        val values = mutableListOf<Any?>()
        for (element in request.arguments ?: emptyList()) {
            val token = CancellationTokenReference.from(element)
            if (token != null) {
                // Keyed on the reference's own id, not on the request's cancellationGuid: the guid
                // cancels the call, the reference cancels this parameter, and two token parameters
                // must be cancellable apart.
                ids.add(token.id)
                cancellationManager.register(token.id, job)
                values.add(ServerCancellationToken(token.id, job))
                continue
            }
            val streamRef = StreamReference.from(element)
            if (streamRef != null) {
                values.add(streamResolver.resolve(streamRef, messageAccessTokenProvider()))
                continue
            }
            values.add(element)
        }
        return ServerMethodArgs(values, json)
    }

    private suspend fun streamBackToServer(request: ServerRequestMessage, streamId: String) {
        val streamHandler = serverStreamHandlers[request.method]
        val handler = serverHandlers[request.method]
        if (streamHandler == null && handler == null) {
            client.logger.warning {
                "No server-method handler registered for streaming request '${request.method}'. " +
                    "Register it with onServerStreamMethod(\"${request.method}\") { ... } or onServerMethod(...)."
            }
            runCatching { client.send(MethodNames.STREAM_COMPLETE_TO_SERVER, listOf(JsonPrimitive(streamId), JsonNull)) }
            return
        }
        try {
            runHandler(request) { args ->
                if (streamHandler != null) {
                    streamHandler(args).collect { item ->
                        client.send(MethodNames.STREAM_ITEM_TO_SERVER, listOf(JsonPrimitive(streamId), JsonValues.encode(item, json)))
                    }
                } else {
                    val result = handler!!(args)
                    client.send(MethodNames.STREAM_ITEM_TO_SERVER, listOf(JsonPrimitive(streamId), encodeResult(result) ?: JsonNull))
                }
            }
            runCatching { client.send(MethodNames.STREAM_COMPLETE_TO_SERVER, listOf(JsonPrimitive(streamId), JsonNull)) }
        } catch (e: CancellationException) {
            if (!currentCoroutineContext().isActive) throw e
            runCatching { client.send(MethodNames.STREAM_COMPLETE_TO_SERVER, listOf(JsonPrimitive(streamId), JsonPrimitive(e.message ?: "cancelled"))) }
        } catch (e: Throwable) {
            client.logger.error(e) { "Streaming handler '${request.method}' failed: ${e.message}" }
            runCatching { client.send(MethodNames.STREAM_COMPLETE_TO_SERVER, listOf(JsonPrimitive(streamId), JsonPrimitive(e.message ?: e.toString()))) }
        }
    }

    // ---------------------------------------------------------------- encoding

    private suspend fun buildClientRequest(method: String, arguments: List<Any?>, genericArguments: List<String>): ClientRequestMessage {
        val token = messageAccessTokenProvider() ?: ""
        val encoded = arguments.map { encodeArgument(it) }
        return ClientRequestMessage(method, encoded, token, genericArguments)
    }

    private fun encodeRequest(request: ClientRequestMessage): JsonElement =
        JsonValues.encode(request, ClientRequestMessage.serializer(), json)

    /** Values with upload bodies are uploaded and replaced by a stream reference; everything else is encoded. */
    private suspend fun encodeArgument(value: Any?): JsonElement {
        val body = StreamReferenceResolver.bodyOf(value)
        if (body != null) return uploadAndReference(body).toJson()
        return JsonValues.encode(value, json)
    }

    private suspend fun encodeResult(result: Any?): JsonElement? {
        if (result == null || result == Unit) return null
        val body = StreamReferenceResolver.bodyOf(result)
        if (body != null) return uploadAndReference(body).toJson()
        return JsonValues.encode(result, json)
    }

    private suspend fun uploadAndReference(body: okhttp3.RequestBody): StreamReference {
        val slot = client.invoke(MethodNames.REQUEST_UPLOAD_SLOT, emptyList())
        val uploadUrl = (slot as? JsonPrimitive)?.takeIf { it.isString }?.content
            ?: throw StreamTransferException("RequestUploadSlot returned no URL")
        streamResolver.upload(uploadUrl, body, messageAccessTokenProvider())
        return StreamReference(uploadUrl)
    }

    private inline fun <R> translateErrors(block: () -> R): R {
        try {
            return block()
        } catch (e: HubInvocationException) {
            throw translate(e)
        }
    }

    private fun translate(e: HubInvocationException): HARRRException {
        if (e is HARRRException) return e
        return HARRRException(HARRRError.parse(e.message ?: ""), e.message ?: "")
    }
}

/** The error reported to the server when it cancels a handler; also thrown by fire-and-forget dispatch. */
public class OperationCancelledException(message: String, cause: Throwable? = null) : SignalRException(message, cause)
